pkg/providers/amifamily/al2.go

```diff
@@ -28,47 +34,61 @@ import (
 
 	"github.com/aws/karpenter-provider-aws/pkg/providers/amifamily/bootstrap"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/ssm"
 )
 
 type AL2 struct {
 	DefaultFamily
 	*Options
 }
 
-// DefaultAMIs returns the AMI name, and Requirements, with an SSM query
-func (a AL2) DefaultAMIs(version string) []DefaultAMIOutput {
-	return []DefaultAMIOutput{
-		{
-			Query: fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2/recommended/image_id", version),
-			Requirements: scheduling.NewRequirements(
-				scheduling.NewRequirement(v1.LabelArchStable, v1.NodeSelectorOpIn, corev1beta1.ArchitectureAmd64),
-				scheduling.NewRequirement(v1beta1.LabelInstanceGPUCount, v1.NodeSelectorOpDoesNotExist),
-				scheduling.NewRequirement(v1beta1.LabelInstanceAcceleratorCount, v1.NodeSelectorOpDoesNotExist),
-			),
-		},
-		{
-			Query: fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2-gpu/recommended/image_id", version),
-			Requirements: scheduling.NewRequirements(
-				scheduling.NewRequirement(v1.LabelArchStable, v1.NodeSelectorOpIn, corev1beta1.ArchitectureAmd64),
-				scheduling.NewRequirement(v1beta1.LabelInstanceGPUCount, v1.NodeSelectorOpExists),
-			),
-		},
-		{
-			Query: fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2-gpu/recommended/image_id", version),
-			Requirements: scheduling.NewRequirements(
-				scheduling.NewRequirement(v1.LabelArchStable, v1.NodeSelectorOpIn, corev1beta1.ArchitectureAmd64),
-				scheduling.NewRequirement(v1beta1.LabelInstanceAcceleratorCount, v1.NodeSelectorOpExists),
-			),
-		},
-		{
-			Query: fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2-%s/recommended/image_id", version, corev1beta1.ArchitectureArm64),
-			Requirements: scheduling.NewRequirements(
-				scheduling.NewRequirement(v1.LabelArchStable, v1.NodeSelectorOpIn, corev1beta1.ArchitectureArm64),
-				scheduling.NewRequirement(v1beta1.LabelInstanceGPUCount, v1.NodeSelectorOpDoesNotExist),
-				scheduling.NewRequirement(v1beta1.LabelInstanceAcceleratorCount, v1.NodeSelectorOpDoesNotExist),
-			),
-		},
+func (a AL2) AMIQuery(ctx context.Context, ssmProvider ssm.Provider, k8sVersion string, amiVersion string) (AMIQuery, error) {
+	query := AMIQuery{
+		Filters: []*ec2.Filter{&ec2.Filter{
+			Name: lo.ToPtr("image-id"),
+		}},
+		KnownRequirements: make(map[string][]scheduling.Requirements),
+	}
+	for rootPath, variants := range map[string][]Variant{
+		fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2", k8sVersion):       []Variant{VariantStandard},
+		fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2-arm64", k8sVersion): []Variant{VariantStandard},
+		fmt.Sprintf("/aws/service/eks/optimized-ami/%s/amazon-linux-2-gpu", k8sVersion):   []Variant{VariantNeuron, VariantNvidia},
+	} {
+		results, err := ssmProvider.List(ctx, rootPath)
+		if err != nil {
+			log.FromContext(ctx).WithValues("path", rootPath).Error(err, "discovering AMIs from ssm")
+			continue
```

Comment:
It's a good question. We should only fail on a subset of paths for a couple of reasons:
- A path is removed (e.g. AL2 no longer publishes GPU AMIs)
- Path formatting has changed, but only for a subset of paths
- We loose API connectivity between calls

In the first two situations, I think the current approach of failing open is correct since the information wouldn't change with a retry (also both are unlikely to change, at least not without warning). IMO the last point could be a good reason to fail closed here. Let's say the API call fails when discovering GPU AMIs. Since we fail open, we'll populate our NodeClass' status with all of the AMIs but the GPU AMI and NodePools referencing that NodeClass will be unable to provision nodes for GPU workloads. This should correct itself in 5 minutes (if the connectivity issue resolves) but we may be able to correct the situation quicker if we retry with exponential backoff on an error.

I'm honestly not sure what the correct approach here would be. My gut is that we should continue to fail open, that way changes to the SSM paths don't cripple this version of Karpenter, requiring **all** users to upgrade. Instead, only users relying on AMIs with the updated paths would need to upgrade to a version of Karpenter that supports the new path.

Reply:
I agree with your assessment. Let's fail open and log.